subsys/dfu/fmfu_fdev/src/fmfu_fdev.c

```diff
@@ -0,0 +1,260 @@
+	if (err != 0) {
+		return err;
+	}
+
+	for (size_t p_offs = offset; p_offs < end; p_offs += buffer_len) {
+		size_t part_len = MIN(buffer_len, (end - p_offs));
+		int err = flash_read(fdev, p_offs, buffer, part_len);
+
+		if (err != 0) {
+			return err;
+		}
+
+		err = mbedtls_sha256_update_ret(&sha256_ctx, buffer, part_len);
+		if (err != 0) {
+			return err;
+		}
+	}
+
+	err = mbedtls_sha256_finish_ret(&sha256_ctx, hash);
+	if (err != 0) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int prevalidate(const COSE_Sign1_Manifest_t *wrapper, const uint8_t *pk,
+		       size_t pk_len, uint8_t *buf, size_t buf_len)
+{
+	return 0;
+}
+
+static int get_blob_len(const uint8_t *segments_buf, size_t buf_len,
+			size_t *segments_len, Segments_t *segments)
+{
+	if (!cbor_decode_Segments(segments_buf, buf_len, segments, NULL)) {
+		return -EINVAL;
+	}
+
+	*segments_len = 0;
+	for (int i = 0; i < segments->_Segments__Segment_count; i++) {
+		*segments_len += segments->_Segments__Segment[i]._Segment_len;
+	}
+
+	return 0;
+}
+
+static int load_segment(const struct device *fdev, size_t seg_size,
+			uint32_t seg_target_addr, uint32_t seg_offset)
+{
+	int err;
+	uint32_t read_addr = seg_offset;
+	size_t bytes_left = seg_size;
+
+	while (bytes_left) {
+		uint32_t read_len = MIN(NRF_FMFU_MODEM_BUFFER_SIZE, bytes_left);
+
+		err = flash_read(fdev, read_addr, buf, read_len);
+		if (err != 0) {
+			LOG_ERR("flash_read failed2: %d", err);
+			return err;
+		}
+
+		err = nrf_fmfu_memory_chunk_write(seg_target_addr, read_len,
+						  buf);
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_memory_chunk_write failed: %d", err);
+			return err;
+		}
+
+		LOG_DBG("wrote chunk: offset 0x%x target addr 0x%x size 0x%x",
+			read_addr, seg_target_addr, read_len);
+
+		seg_target_addr += read_len;
+		bytes_left -= read_len;
+		read_addr += read_len;
+	}
+
+	return 0;
+}
+
+static int load_segments(Segments_t *seg, size_t blob_offset)
+{
+	int err;
+	size_t prev_segments_len = 0;
+
+	LOG_INF("Writing %d segments", seg->_Segments__Segment_count);
+
+	for (int i = 0; i < seg->_Segments__Segment_count; i++) {
+		size_t seg_size = seg->_Segments__Segment[i]._Segment_len;
+		uint32_t seg_addr =
+			seg->_Segments__Segment[i]._Segment_target_addr;
+		uint32_t read_addr = blob_offset + prev_segments_len;
+
+		err = nrf_fmfu_transfer_start();
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_transfer_start failed1: %d", err);
+			return err;
+		}
+
+		err = load_segment(fdev, seg_size, seg_addr, read_addr);
+		if (err != 0) {
+			LOG_ERR("load_segment failed: %d", err);
+			return err;
+		}
+
+		err = nrf_fmfu_transfer_end();
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_transfer_end failed: %d", err);
+			return err;
+		}
+
+		prev_segments_len += seg_size;
+
+		LOG_INF("Segment %d written. Target addr: 0x%x, size: 0%x",
+			i, seg_addr, seg_size);
+	}
+
+	LOG_INF("FMFU finished");
+
+	return 0;
+}
+
+int fmfu_fdev_init(uint8_t *buf_in, size_t buf_len_in,
```

Comment:
I'm not sure storing the buf and fdev as static vars is the best solution. Especially now that fmfu_fdev_load is the only other function.

It gives problems with reentancy, and it's less clear when the buf is not used any more.

Reply:
This is not trivial, lets take an offline discussion.

Reply:
I will remove _init and do it all from _load.